### ISSUE TITLE
Not move primary key condition to prewhere to optimize left table filter marks when cross join

### ISSUE THREAD

**Issue by LiuYangkuan** [ClickHouse/ClickHouse#50399]
**Describe the situation**
Our cross join query encounters performance degradation from 20.8 to latest version.
We found that there are two different query plans between these two versions.

**How to reproduce**
* Which ClickHouse server version to use 
20.8 and 23.4

```
CREATE TABLE t1 (id Int64, create_time DateTime) ENGINE = MergeTree ORDER BY id  AS SELECT number, toDateTime(number + 1600000000) from numbers(10000000) settings min_insert_block_size_rows=100000;
CREATE TABLE t2 (delete_time DateTime) ENGINE = MergeTree ORDER BY delete_time AS SELECT toDateTime(number + 1610000000) from numbers(100);
```
* Queries to run that lead to slow performance
```
SELECT id, delete_time FROM t1
 CROSS JOIN (
    SELECT delete_time
    FROM t2 WHERE delete_time > '2023-05-30 00:00:00'
) AS d WHERE create_time < delete_time AND id IN (10000001)
```
We can explain plan indexes=1 above query sql, and result is:
```
Expression ((Projection + Before ORDER BY))
  Filter (WHERE)
    Join (JOIN FillRightFirst)
      Expression (Before JOIN)
        ReadFromMergeTree (default.t1)
        Indexes:
          PrimaryKey
            Condition: true
            Parts: 3/3
            Granules: 1222/1222
      Expression ((Joined actions + (Rename joined columns + (Projection + Before ORDER BY))))
        Filter (WHERE)
          ReadFromMergeTree (default.t2)
          Indexes:
            PrimaryKey
              Keys:
                delete_time
              Condition: (delete_time in [1685404801, +Inf))
              Parts: 0/1
              Granules: 0/1
```

**Expected performance**

The query optimizer doesn't move primary key `id` condition as `prewhere` condition to filter table `t1` read marks 

**Additional context**

**Comment by davenger** [comment]
Looks like the main issue here is with ```where id ...``` condition not being pushed down to reading t1 before join.
When I do it "manually" the index of t1 is being used:

```
EXPLAIN indexes=1 SELECT id, delete_time FROM (SELECT * FROM t1 WHERE id IN (10000001)) as a
 CROSS JOIN (
    SELECT delete_time
    FROM t2 WHERE delete_time > '2023-05-30 00:00:00'
) AS d WHERE create_time < delete_time

┌─explain────────────────────────────────────────────────────────────────────────────────────────┐
│ Expression ((Projection + Before ORDER BY))                                                    │
│   Filter (WHERE)                                                                               │
│     Join (JOIN FillRightFirst)                                                                 │
│       Expression ((Before JOIN + (Projection + Before ORDER BY)))                              │
│         ReadFromMergeTree (test_50399.t1)                                                      │
│         Indexes:                                                                               │
│           PrimaryKey                                                                           │
│             Keys:                                                                              │
│               id                                                                               │
│             Condition: (id in 1-element set)                                                   │
│             Parts: 0/3                                                                         │
│             Granules: 0/1200                                                                   │
│       Expression ((Joined actions + (Rename joined columns + (Projection + Before ORDER BY)))) │
│         Filter (WHERE)                                                                         │
│           ReadFromMergeTree (test_50399.t2)                                                    │
│           Indexes:                                                                             │
│             PrimaryKey                                                                         │
│               Keys:                                                                            │
│                 delete_time                                                                    │
│               Condition: (delete_time in [1685397601, +Inf))                                   │
│               Parts: 0/1                                                                       │
│               Granules: 0/1                                                                    │
└────────────────────────────────────────────────────────────────────────────────────────────────┘

```

**Comment by hanfei1991** [comment]
seems filter push down does not consider cross join: https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/QueryPlan/Optimizations/filterPushDown.cpp#L328

@KochetovNicolai can we push down through cross join safely?

**Comment by den-crane** [comment]
>@KochetovNicolai can we push down through cross join safely?

AFIAK it's more complicated. 
WHERE must work after JOIN, and **cross_join** can produce NULLs(default values) and the only pushdown of `where some_col is [not] null / =[!=]0` may produce incorrect results, it can be solved by a secondary after join filtering.

**Comment by hanfei1991** [comment]
@den-crane why cross join will produce NULL ? In my mind it's an inner join without any conditions.

**Comment by den-crane** [comment]
right, I was thinking about full outer join

**Comment by alexey-milovidov** [comment]
Reopened due to failed tests.

**Comment by LiuYangkuan** [comment]
If filter condition of right table not using subquery, then the query plan is also not push down filter to right table.
For example:
```
EXPLAIN PLAN indexes=1
SELECT id, delete_time FROM t1
 CROSS JOIN t2 AS d
 WHERE create_time < delete_time AND id IN (10000001) AND delete_time > '2023-05-30 00:00:00'
┌─explain────────────────────────────────────────────────────────────────────────────────────────┐
│ Expression ((Projection + Before ORDER BY))                                                    │
│   Filter (WHERE)                                                                               │
│     Join (JOIN FillRightFirst)                                                                 │
│       Filter (( + Before JOIN))                                                                │
│         ReadFromMergeTree (default.t1)                                                         │
│         Indexes:                                                                               │
│           PrimaryKey                                                                           │
│             Keys:                                                                              │
│               id                                                                               │
│             Condition: (id in 1-element set)                                                   │
│             Parts: 0/5                                                                         │
│             Granules: 0/1218                                                                   │
│       Expression ((Joined actions + (Rename joined columns + (Projection + Before ORDER BY)))) │
│         ReadFromMergeTree (default.t2)                                                         │
│         Indexes:                                                                               │
│           PrimaryKey                                                                           │
│             Condition: true                                                                    │
│             Parts: 3/3                                                                         │
│             Granules: 105/105                                                                  │
└────────────────────────────────────────────────────────────────────────────────────────────────┘
```
In fact, the right table also need read any granules of parts!

**Comment by SaltTan** [comment]
It seems fixed in 23.7 https://fiddle.clickhouse.com/f59134f2-41ba-475e-ba0d-4fd8eba99aaa
